src/router.rs

```diff
@@ -190,18 +208,26 @@ impl Router {
             url = Url::from_generic_url(generic_url).unwrap();
         }
 
-        self.recognize(&req.method, &path).and(
-            Some(IronError::new(TrailingSlash,
+        self.recognize(&req.method, &path).ok().and(
+            Some(IronError::new(RouterError::TrailingSlash,
                                 (status::MovedPermanently, Redirect(url))))
         )
     }
 
     fn handle_method(&self, req: &mut Request, path: &str) -> Option<IronResult<Response>> {
-        if let Some(matched) = self.recognize(&req.method, path) {
-            req.extensions.insert::<Router>(matched.params);
-            req.extensions.insert::<RouterInner>(self.inner.clone());
-            Some(matched.handler.handle(req))
-        } else { self.redirect_slash(req).and_then(|redirect| Some(Err(redirect))) }
+        match self.recognize(&req.method, path) {
+            Ok(matched) => {
+                req.extensions.insert::<Router>(matched.params);
+                req.extensions.insert::<RouterInner>(self.inner.clone());
+                Some(matched.handler.handle(req))
+            },
+            Err(e) => {
+                if e == RouterError::MethodNotAllowed {
```

Comment:
You can use `Err(RouterError::MethodNotAllowed)` as independent match arm. That would be less indentation.